File: src/github_tools.js
import { Octokit } from "@octokit/rest";

let octokit;

function getOctokit() {
  if (!octokit) {
    const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
    if (!GITHUB_TOKEN) {
      throw new Error("GITHUB_TOKEN environment variable is not set.");
    }
    console.log("[github_tools] Initializing Octokit with GITHUB_TOKEN");
    octokit = new Octokit({ auth: GITHUB_TOKEN });
  }
  return octokit;
}

/**
 * Creates a new branch from a base branch.
 * @param {string} owner - The repository owner.
 * @param {string} repo - The repository name.
 * @param {string} branchName - The name of the new branch.
 * @param {string} baseBranch - The branch to create from (e.g., 'main').
 * @returns {Promise<void>}
 */
async function createBranch(owner, repo, branchName, baseBranch = 'main') {
  const kit = getOctokit();
  try {
    const { data: baseBranchRef } = await kit.git.getRef({
      owner,
      repo,
      ref: `heads/${baseBranch}`,
    });

    await kit.git.createRef({
      owner,
      repo,
      ref: `refs/heads/${branchName}`,
      sha: baseBranchRef.object.sha,
    });
    console.log(`Branch '${branchName}' created successfully.`);
  } catch (error) {
    if (error.status === 422 && error.message.includes("Reference already exists")) {
        console.warn(`Branch '${branchName}' already exists. Skipping creation.`);
    } else {
        console.error(`Error creating branch '${branchName}':`, error);
        throw error;
    }
  }
}

/**
 * Creates a pull request.
 * @param {string} owner - The repository owner.
 * @param {string} repo - The repository name.
 * @param {string} title - The title of the pull request.
 * @param {string} head - The source branch for the pull request.
 * @param {string} base - The target branch for the pull request.
 * @param {string} body - The body/description of the pull request.
 * @returns {Promise<object>} - The created pull request object.
 */
async function createPullRequest(owner, repo, title, head, base, body) {
  const kit = getOctokit();
  try {
    const { data: pullRequest } = await kit.pulls.create({
      owner,
      repo,
      title,
      head,
      base,
      body,
    });
    console.log(`Pull request created: ${pullRequest.html_url}`);
    return pullRequest;
  } catch (error) {
    console.error("Error creating pull request:", error.message);
    if (error.status === 422 && error.errors && error.errors.some(e => e.message.includes("A pull request already exists"))) {
        console.warn(`A pull request for branch '${head}' already exists.`);
        const { data: existingPRs } = await kit.pulls.list({
            owner,
            repo,
            head: `${owner}:${head}`,
            base,
            state: 'open',
        });
        if (existingPRs.length > 0) {
            console.log(`Found existing PR: ${existingPRs[0].html_url}`);
            return existingPRs[0];
        }
    }
    throw error;
  }
}

/**
 * Gets the content of a file from the repository.
 * @param {string} owner - The repository owner.
 * @param {string} repo - The repository name.
 * @param {string} path - The path to the file.
 * @param {string} [ref] - The branch, tag, or commit SHA. Defaults to the default branch.
 * @returns {Promise<string>} - The decoded content of the file.
 */
async function getFileContent(owner, repo, path, ref) {
    const kit = getOctokit();
    try {
        const { data } = await kit.repos.getContent({
            owner,
            repo,
            path,
            ref,
        });

        if (data.type !== 'file') {
            throw new Error(`Path '${path}' is not a file.`);
        }

        return Buffer.from(data.content, 'base64').toString('utf-8');
    } catch (error) {
        console.error(`Error getting file content for '${path}':`, error);
        throw error;
    }
}

/**
 * Commits a file to a branch.
 * @param {string} owner
 * @param {string} repo
 * @param {string} branch
 * @param {string} path
 * @param {string} content
 * @param {string} message
 * @returns {Promise<object>}
 */
async function commitFile(owner, repo, branch, path, content, message) {
    const kit = getOctokit();
    try {
        let currentFileSha;
        try {
            const { data: existingFile } = await kit.repos.getContent({
                owner,
                repo,
                path,
                ref: branch,
            });
            currentFileSha = existingFile.sha;
        } catch (error) {
            if (error.status !== 404) {
                throw error;
            }
        }

        const { data: commitData } = await kit.repos.createOrUpdateFileContents({
            owner,
            repo,
            path,
            message,
            content: Buffer.from(content).toString('base64'),
            branch,
            sha: currentFileSha,
        });

        console.log(`Successfully committed '${path}' to branch '${branch}'. Commit SHA: ${commitData.commit.sha}`);
        return commitData;
    } catch (error) {
        console.error(`Error committing file '${path}':`, error);
        throw error;
    }
}


export {
  createBranch,
  createPullRequest,
  getFileContent,
  commitFile,
};

File: py/github_tools.py
# This file will contain functions to interact with the GitHub API.

import requests
import os
import json
from urllib.parse import urljoin

GITHUB_TOKEN = os.getenv('GITHUB_TOKEN')
HEADERS = {
    'Authorization': f'token {GITHUB_TOKEN}',
    'Accept': 'application/vnd.github.v3+json',
}
BASE_URL = 'https://api.github.com'

def create_pull_request(repo, title, head, base, body):
    """
    Create a pull request on GitHub.
    
    Args:
        repo: Repository in format 'owner/repo'
        title: Pull request title
        head: Source branch name
        base: Target branch name
        body: Pull request description
    
    Returns:
        dict: Response from GitHub API containing PR details (number, html_url, etc.)
    
    Raises:
        Exception: If the API request fails
    """
    if not GITHUB_TOKEN:
        raise ValueError("GITHUB_TOKEN environment variable is not set")
    
    if '/' not in repo:
        raise ValueError(f"Invalid repo format: {repo}. Expected 'owner/repo'")
    
    owner, repo_name = repo.split('/', 1)
    
    url = f'{BASE_URL}/repos/{owner}/{repo_name}/pulls'
    
    payload = {
        'title': title,
        'head': head,
        'base': base,
        'body': body,
        'draft': False
    }
    
    try:
        response = requests.post(url, headers=HEADERS, json=payload)
        
        if response.status_code == 201:
            return response.json()
        elif response.status_code == 422:
            # PR might already exist or validation error
            error_data = response.json()
            raise Exception(f"GitHub API 422 error: {error_data.get('message', 'Unknown error')}")
        else:
            error_data = response.json() if response.text else {}
            raise Exception(f"GitHub API {response.status_code}: {error_data.get('message', response.text)}")
    
    except requests.RequestException as e:
        raise Exception(f"Failed to create pull request: {str(e)}")


def create_branch(repo, branch_name, source_branch='main'):
    """
    Create a new branch in a GitHub repository.
    
    Args:
        repo: Repository in format 'owner/repo'
        branch_name: Name of the new branch to create
        source_branch: Name of the branch to create from (default: 'main')
    
    Returns:
        dict: Response from GitHub API containing branch details
    
    Raises:
        Exception: If the API request fails
    """
    if not GITHUB_TOKEN:
        raise ValueError("GITHUB_TOKEN environment variable is not set")
    
    if '/' not in repo:
        raise ValueError(f"Invalid repo format: {repo}. Expected 'owner/repo'")
    
    owner, repo_name = repo.split('/', 1)
    
    # First, get the SHA of the source branch
    url = f'{BASE_URL}/repos/{owner}/{repo_name}/git/refs/heads/{source_branch}'
    
    try:
        response = requests.get(url, headers=HEADERS)
        
        if response.status_code != 200:
            raise Exception(f"Failed to get source branch {source_branch}: {response.status_code}")
        
        source_sha = response.json()['object']['sha']
        
        # Create new branch pointing to the same commit
        create_url = f'{BASE_URL}/repos/{owner}/{repo_name}/git/refs'
        payload = {
            'ref': f'refs/heads/{branch_name}',
            'sha': source_sha
        }
        
        response = requests.post(create_url, headers=HEADERS, json=payload)
        
        if response.status_code == 201:
            return response.json()
        else:
            error_data = response.json() if response.text else {}
            raise Exception(f"GitHub API {response.status_code}: {error_data.get('message', response.text)}")
    
    except requests.RequestException as e:
        raise Exception(f"Failed to create branch: {str(e)}")


def commit_file(repo, branch, file_path, content, commit_message):
    """
    Create or update a file on a branch with a commit.
    
    Args:
        repo: Repository in format 'owner/repo'
        branch: Target branch name
        file_path: Path to the file in the repository
        content: Content to write to the file
        commit_message: Git commit message
    
    Returns:
        dict: Response from GitHub API containing commit details
    
    Raises:
        Exception: If the API request fails
    """
    if not GITHUB_TOKEN:
        raise ValueError("GITHUB_TOKEN environment variable is not set")
    
    if '/' not in repo:
        raise ValueError(f"Invalid repo format: {repo}. Expected 'owner/repo'")
    
    owner, repo_name = repo.split('/', 1)
    
    # First check if file exists
    url = f'{BASE_URL}/repos/{owner}/{repo_name}/contents/{file_path}'
    
    import base64
    
    try:
        response = requests.get(url, headers=HEADERS, params={'ref': branch})
        
        sha = None
        if response.status_code == 200:
            sha = response.json()['sha']
        
        # Encode content to base64
        encoded_content = base64.b64encode(content.encode('utf-8')).decode('utf-8')
        
        payload = {
            'message': commit_message,
            'content': encoded_content,
            'branch': branch
        }
        
        if sha:
            payload['sha'] = sha
        
        response = requests.put(url, headers=HEADERS, json=payload)
        
        if response.status_code in [200, 201]:
            return response.json()
        else:
            error_data = response.json() if response.text else {}
            raise Exception(f"GitHub API {response.status_code}: {error_data.get('message', response.text)}")
    
    except requests.RequestException as e:
        raise Exception(f"Failed to commit file: {str(e)}")
